examples/chat.rs

```diff
@@ -124,25 +126,21 @@ fn run(socket: UdpSocket, rx: Receiver<Rtc>) -> Result<(), RtcError> {
             clients.push(client);
         }
 
-        // Poll all clients, and get propagated events as a result.
-        let to_propagate: Vec<_> = clients.iter_mut().map(|c| c.poll_output(&socket)).collect();
-        let timeouts: Vec<_> = to_propagate.iter().filter_map(|p| p.as_timeout()).collect();
-
-        // We keep propagating client events until all clients respond with a timeout.
-        if to_propagate.len() > timeouts.len() {
-            propagate(&mut clients, to_propagate);
-            // Start over to propagate more client data until all are timeouts.
-            continue;
-        }
-
-        // Timeout in case we have no clients. We can't wait forever since we need to keep
-        // polling the spawn_new_clients to discover a client.
-        fn default_timeout() -> Instant {
-            Instant::now() + Duration::from_millis(100)
-        }
-
-        // All poll_output resulted in timeouts, figure out the shortest timeout.
-        let timeout = timeouts.into_iter().min().unwrap_or_else(default_timeout);
+        let timeout = loop {
+            let propagated = propagation.pop_front().unwrap_or(Propagated::Noop);
+            let mut timeout = Instant::now() + Duration::from_millis(100);
+            let mut timeouts = 0;
+            for client in clients.iter_mut() {
+                let t = propagate_one(&propagated, client, &mut propagation, &socket);
+                timeout = timeout.min(t);
+                timeouts += 1
+            }
+            if propagation.is_empty() && timeouts == clients.len() {
+                // we have progated everything in the propagation queue and all
+                // clients have returned a timeout
+                break timeout;
+            }
+        };
```

Comment:
This is a bit messy with how that timeout is created every loop, and why is the timeout in this loop at all there is no other timeout that it can end with. Please revise.

Reply:
Right. There's a second loop.

I don't like that second loop. Polling output has nothing to do with propagation, so it shouldn't be inside something called `propagate_one`.

This whole PR is taking the pattern too far away from the documented patterns we have in the README.md etc.

Reply:
I would start over, think "minimal changes" and make sure it follows the README.md in terms of structure.

Reply:
If I am not wrong, the contract is to 
- send input
- poll all output until you get a timeout

At this point there is no change in state until either time passes (timeout) or there is new input.

propagate_one does exactly that

Reply:
the example in the readme does the same, but it's just inverted.

- first it polls everything until there is a timeout
- then reads an input from the network and sends it through

From a state-machine perspective there is no difference.

However in a typical application you "wait" the network and then react, so I find this sequence more natural.